### PR TITLE
[DEPLOY-512]: Adds L2EP Foundry Tests

### PR DESCRIPTION
**Pull request by chris-de-leon-cll**
### Overview
- This PR adds Foundry test cases for the L2EP contracts (previously we were relying on Hardhat tests)
- This PR is one of a few related to the L2EP modernization effort - the effort consists of the following components:
  - Migrate old L2EP hardhat tests to Foundry tests
  - Refactor the L2EP contracts such that:
    - They comply with the style guide
    - They are more extensible to other chains
    - They include the gas optimizations and other QoL improvements made for Scroll

### Important Note(s)
- The Foundry tests have been written such that they test the same scenarios as the Hardhat tests - the intention behind this was to reduce overhead and avoid having to reinvent the wheel
- The Scroll and Optimism Foundry tests are almost identical - the only changes include the imports and contract names. Arbitrum is slightly different, and the test cases are not identical to Scroll and Optimism
- If we expect future L2EP solutions will be similar to that of Scroll and Optimism, we may want to consider revising the these test cases such that we don't need to copy-paste the same code each time we integrate a new chain
- The Hardhat testing suite included some test scenarios that manually measured gas usage. These have also been migrated to Foundry, but they might be unnecessary since Foundry already collects gas usage metrics. Any clarity on whether we should still include gas measuring tests in Foundry would be much appreciated!

#### Here's a breakdown of some of the primary changes included in this PR:
- [x] Adds Foundry configs for L2EP
- [x] Adds ScrollValidator test
- [x] Adds ScrollSequencerUptimeFeed test
- [x] Adds ScrollCrossDomainGovernor test
- [x] Adds ScrollCrossDomainForwarder test
- [x] Adds OptimismValidator test
- [x] Adds OptimismSequencerUptimeFeed test
- [x] Adds OptimismCrossDomainGovernor test
- [x] Adds OptimismCrossDomainForwarder test
- [x] Adds ArbitrumValidator test
- [x] Adds ArbitrtumSequencerUptimeFeed test
- [x] Adds ArbitrumDomainGovernor test
- [x] Adds ArbitrumDomainForwarder test
- [x] Adds additional testing mocks 
- [x] Adds documentation (see the README.md file in `contracts/src/v0.8/l2ep/README.md`)
- [x] Ensures Foundry tests comply with style guide

### CODE REVIEW

File: contracts/src/v0.8/l2ep/README.md [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+A full list of flags for `forge test` can be found [here](https://book.getfoundry.sh/reference/forge/forge-test).
+
+#### Coverage
+
+First ensure that the correct files are being evaluated. For example, if only
+v1 contracts are, being evaluated then temporarily change the L2EP profile in
+`./foundry.toml`.
+
+```sh
+forge coverage --ir-minimum
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/README.md:122]: Is the --ir-minimum required? Does that not make it very very slow?

File: contracts/src/v0.8/l2ep/test/mocks/optimism/MockOVMCrossDomainMessenger.sol [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+
+  function xDomainMessageSender() external view override returns (address) {
+    return s_mockMessageSender;
+  }
+
+  function _setMockMessageSender(address sender) external {
+    s_mockMessageSender = sender;
+  }
+
+  /********************
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/mocks/optimism/MockOVMCrossDomainMessenger.sol:23]: nit: probably not required to have a header here

File: contracts/src/v0.8/l2ep/test/v1_0_0/Example.t.sol [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+
+import {Test} from "forge-std/Test.sol";
+
+// Use the following command to run this test file:
+//
+//  FOUNDRY_PROFILE=l2ep forge test -vvv --mp ./src/v0.8/l2ep/test/v1_0_0/Example.t.sol
+//
+contract ExampleTest is Test {
+  function test_example() public {
+    assertEq(true, true);
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/Example.t.sol:12]: to be removed?

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol [hunk 4]
```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import {ArbitrumCrossDomainForwarder} from "../../../dev/arbitrum/ArbitrumCrossDomainForwarder.sol";
+import {Greeter} from "../../../../../v0.8/tests/Greeter.sol";
+import {L2EPTest} from "../L2EPTest.t.sol";
+
+// Use this command from the /contracts directory to run this test file:
+//
+//  FOUNDRY_PROFILE=l2ep forge test -vvv --match-path ./src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:10]: Probably not required to start files with this comment. I generally use this syntax btw
```
forge test -vvv --match-contract ArbitrumCrossDomainForwarderTest
```
which is a lot easier to modify between different calls. You also don't even have to type the complete name, so 
```
forge test -vvv --match-contract ArbitrumCro
```

would work as well

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol [hunk 5]
```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:2]: Generally we lock tests to a given version

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+import {Greeter} from "../../../../../v0.8/tests/Greeter.sol";
+import {L2EPTest} from "../L2EPTest.t.sol";
+
+// Use this command from the /contracts directory to run this test file:
+//
+//  FOUNDRY_PROFILE=l2ep forge test -vvv --match-path ./src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol
+//
+contract ArbitrumCrossDomainForwarderTest is L2EPTest {
+  /// Helper variables
+  address internal s_strangerAddr = vm.addr(0x1);
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:14]: These could be moved to L2EPTest correct?

Comment (RensR) [reply contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:14]: In CCIP we have OWNER and STRANGER as the 2 default callers

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol [hunk 7]
```diff
@@ -0,0 +1,191 @@
...
+  function setUp() public {
+    // Deploys contracts
+    vm.startPrank(s_l1OwnerAddr, s_l1OwnerAddr);
+    s_arbitrumCrossDomainForwarder = new ArbitrumCrossDomainForwarder(s_l1OwnerAddr);
+    s_greeter = new Greeter(address(s_arbitrumCrossDomainForwarder));
+    vm.stopPrank();
+  }
+}
+
+contract Constructor is ArbitrumCrossDomainForwarderTest {
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:37]: I would name this <contract>_<function> so the gas snapshot and CLI are more readable

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol [hunk 8]
```diff
@@ -0,0 +1,191 @@
...
+  }
+}
+
+contract Forward is ArbitrumCrossDomainForwarderTest {
+  /// @notice it should not be callable by unknown address
+  function test_NotCallableByUnknownAddress() public {
+    vm.startPrank(s_strangerAddr, s_strangerAddr);
+    vm.expectRevert("Sender is not the L2 messenger");
+    s_arbitrumCrossDomainForwarder.forward(address(s_greeter), abi.encode(""));
+    vm.stopPrank();
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:65]: You don't have to stop pranks

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol [hunk 9]
```diff
@@ -0,0 +1,494 @@
...
+    uint256 timestamp = s_arbitrumSequencerUptimeFeed.latestTimestamp();
+    assertEq(s_arbitrumSequencerUptimeFeed.latestAnswer(), 0);
+
+    // Defines helper variables for measuring gas usage
+    uint256 expectedGasUsed;
+    uint256 gasStart;
+    uint256 gasFinal;
+
+    // measures gas used for no update
+    expectedGasUsed = 5507; // TODO: used to be 28300
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol:292]: That sounds like a good improvement :p

Comment (RensR) [reply contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol:292]: Generally you don't need specific gas tests in Foundry, as all tests will be part of the snapshot anyway. So you will always have full gas coverage on all tests. You can however still do more accurate gas measurements with 

```
vm.pauseGasMetering();
vm.resumeGasMetering();
```

those calls stop and start gas measurements for when you are doing setup calls/logic in the test. You then only turn on measuring for the single call you want to measure. 

Take this test as example


```
  function testTransmitSuccess_gas() public {
    vm.pauseGasMetering();
    bytes32[3] memory reportContext = [s_configDigest, s_configDigest, s_configDigest];

    changePrank(s_valid_transmitters[0]);
    vm.resumeGasMetering();
    s_OCR2Base.transmit(reportContext, REPORT, s_rs, s_ss, s_rawVs);
  }
```


You could even load storage vars into memory before the call to be more exact

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol [hunk 10]
```diff
@@ -0,0 +1,494 @@
...
+    gasFinal = gasleft();
+
+    // Checks that gas usage is within expected range
+    assertGasUsageIsCloseTo(expectedGasUsed, gasStart, gasFinal, GAS_USED_DEVIATION);
+
+    // Resets msg.sender and tx.origin
+    vm.stopPrank();
+  }
+
+  /// @notice it should consume a known amount of gas for getTimestamp() @skip-coverage
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol:469]: Can remove the HH skip cov comments

**Comment by RensR** [comment]
You probably want to add l2ep to the Foundry-using-projects list at the top of `contracts/GNUmakefile` and in CI `.github/workflows/solidity-foundry.yml` line 35. This will make CI run your new tests. You should also create a snapshot with 

```
make snapshot
```

with `export FOUNDRY_PROFILE=l2ep` set. With that set you can run `make snapshot` and `make wrappers` for just l2ep, so it's very quick.

File: contracts/remappings.txt [hunk 11]
```diff
@@ -1,6 +1,7 @@
-ds-test/=foundry-lib/forge-std/lib/ds-test/src
-forge-std/=foundry-lib/forge-std/src
+ds-test/=foundry-lib/forge-std/lib/ds-test/src/
+forge-std/=foundry-lib/forge-std/src/
```
[the remark this answers is not in the document]
Comment (chris-de-leon-cll) [reply contracts/remappings.txt:2]: When I was setting up my environment, I found that if I excluded the slash, my editor would take

```sol
import forge-std/Test.sol
```

and expand it to

```sol
import foundry-lib/forge-std/srcTest.sol
```

causing it to resolve the path incorrectly

File: contracts/src/v0.8/l2ep/README.md (hunk 1 above)
Comment (chris-de-leon-cll) [reply contracts/src/v0.8/l2ep/README.md:122]: I didn't notice any significant slow down with the flag, but even so you're correct it is not required - I'll remove this from the README!

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol [hunk 12]
```diff
@@ -0,0 +1,191 @@
...
+    s_arbitrumCrossDomainForwarder.transferL1Ownership(s_strangerAddr);
+
+    // Resets msg.sender and tx.origin
+    vm.stopPrank();
+  }
+
+  /// @notice it should be callable by current L1 owner to zero address
+  function test_CallableByL1OwnerOrZeroAddress() public {
+    // Sets msg.sender and tx.origin
+    vm.startPrank(s_crossDomainMessengerAddr, s_crossDomainMessengerAddr);
```
[the remark this answers is not in the document]
Comment (chris-de-leon-cll) [reply contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:143]: I double checked the places I was using it and simplified each of the calls - `tx.origin` is only used for the SequencerUptimeFeed tests

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol (hunk 7 above)
Comment (chris-de-leon-cll) [reply contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol:37]: Ah I see what you mean 

It looks like `forge snapshot` doesn't include the path to the file or any information about the parent contract name, so you just get a bunch of duplicate lines

If I change the contract name to _, I run into an error, so for now I've prefixed the tests with a more specific name - lmk what you think!

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol [hunk 13]
```diff
@@ -0,0 +1,191 @@
...
+    vm.stopPrank();
+  }
+
+  /// @notice it should be callable by current L1 owner
+  function test_CallableByL1Owner() public {
+    // Sets msg.sender and tx.origin
+    vm.startPrank(s_crossDomainMessengerAddr, s_crossDomainMessengerAddr);
+
+    // Defines the cross domain message to send
+    vm.expectEmit(false, false, false, true);
```
[the remark this answers is not in the document]
Comment (chris-de-leon-cll) [reply contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol]: I revisited each of the areas where `vm.expectEmit` is used and modified them accordingly, and there were only a couple places where the extended version was needed - good call here!

File: contracts/src/v0.8/l2ep/test/v1_0_0/L2EPTest.t.sol [hunk 14]
```diff
@@ -0,0 +1,75 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import {MultiSend} from "../../../../v0.8/vendor/MultiSend.sol";
```
[the remark this answers is not in the document]
Comment (chris-de-leon-cll) [reply contracts/src/v0.8/l2ep/test/v1_0_0/L2EPTest.t.sol]: Made the change to this file and in a few other places as well - good find!

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol (hunk 7 above)
Comment (RensR) [reply contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumCrossDomainForwarder.t.sol]: Oh that was a formatting issue. I meant to type <contract>_<function> as name

File: contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol [hunk 15]
```diff
@@ -1,24 +1,19 @@
...
 contract ArbitrumSequencerUptimeFeedTest is L2EPTest {
   /// Constants
   uint256 internal constant GAS_USED_DEVIATION = 100;
 
-  /// Helper variables
-  address internal s_mockL1OwnerAddr = vm.addr(0x1);
-  address internal s_strangerAddr = vm.addr(0x2);
+  /// Helper variable(s)
+  address internal s_mockL1OwnerAddr = vm.addr(0x2);
   address internal s_deployerAddr = vm.addr(0x3);
```
Comment (RensR) [inline contracts/src/v0.8/l2ep/test/v1_0_0/arbitrum/ArbitrumSequencerUptimeFeed.t.sol:17]: Deployer address also seems to be useful in the higher level one

**Review by RensR — APPROVED** [review]

**Review by RensR — APPROVED** [review]

**Pull request merged**